### PR TITLE
fix: testground run with --wait exits with code = 0 on some failures

### PR DESCRIPTION
**Pull request by laurentsenta**
Fixes https://github.com/testground/testground/issues/1349
Contributes to https://github.com/testground/testground/issues/1304

## Context:

Testground defines a `task.Task` object with an anonymous Result field. The content of this field depends on the builder or runner used, for example:

- if the task is Build => Result contains an array of artifacts (probably docker images),
- if the task is Run, with a local_exec runner => Result contains nil,
- if the Task is Run, with a k8s or local docker runner => Result contains a struct with different Outcomes field.

That's fine for storage / serialization (when used as a wire format), but any piece of code that "receive" a task object needs to interpret the `Result` field by knowing about the runners. There are hints of the code required in the `daemon/tasks.go` code, see related PR: https://github.com/testground/testground/pull/1321. It create cyclic dependencies and bugs (because not all parts of the code use the same interpretation, see issue #1349).

Later it would be nice to rethink this data model.

## Tasks

- [x] introduce a `data.DecodeTaskOutcome` function, (a non-intrusive way to make sure we have ONE reusable way to compute task outcomes),
- [x] make sure running a job with `--wait` will fail with exit code != 0 if the outcome is not a success.
- [x] Return with an exit code 3 if the command failed during collect.
- [x] Update tests: now that we check for actual outcomes, most test are failing

### CODE REVIEW

**Comment by laurentsenta** [comment]
I'll revisit https://github.com/testground/testground/pull/1321 when we approve & merge this.

**Comment by laurentsenta** [comment]
(temporarily adding the fix from https://github.com/testground/testground/pull/1353 to make sure the CI is passing)

File: pkg/data/result_test.go [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+	}
+	r1 := DecodeRunnerResult(result1)
+	assert.NotNil(t, r1)
+
+	result2 := [2]string{"artifact", "artifact2"}
+	r2 := DecodeRunnerResult(result2)
+	assert.NotNil(t, r2)
+}
+
+func TestDecodeTaskOutcome(t *testing.T) {
```
Comment (brdji) [inline pkg/data/result_test.go:29]: I'd rather split this large test into several smaller unit tests, each one testing a specific outcome (and the decoding of the outcome).
E.g. `TestDecodeGenericUnknownOutcome`, `TestDecodeGenericSuccessOutcome`, etc.

Comment (laurentsenta) [reply pkg/data/result_test.go]: Fixed!

**Pull request merged**